src/editor/codemirror/language-server/signatureHelp.ts

```diff
@@ -0,0 +1,171 @@
+      }
+    }
+    return state;
+  },
+  provide: (f) => showTooltip.from(f, (val) => val.tooltip),
+});
+
+export class SignatureHelpView
+  extends BaseLanguageServerView
+  implements PluginValue
+{
+  update({ docChanged, selectionSet, changes, transactions }: ViewUpdate) {
+    if (
+      (docChanged || selectionSet) &&
+      this.view.state.field(signatureHelpTooltipField).tooltip
+    ) {
+      this.triggerSignatureHelpRequest();
+    } else if (docChanged) {
+      const last = transactions[transactions.length - 1];
+      if (last.isUserEvent("input.type")) {
+        last.changes.iterChanges((_fromA, _toA, _fromB, _toB, inserted) => {
+          if (inserted.eq(Text.of(["("])) || inserted.eq(Text.of(["()"]))) {
+            this.triggerSignatureHelpRequest();
+          }
+        });
+      }
+    }
+  }
+
+  async triggerSignatureHelpRequest() {
+    const pos = this.view.state.selection.main.from;
+    const params: SignatureHelpParams = {
+      textDocument: { uri: this.uri },
+      position: offsetToPosition(this.view.state.doc, pos),
+    };
+    try {
+      const result = await this.client.connection.sendRequest(
+        SignatureHelpRequest.type,
+        params
+      );
+      this.view.dispatch({
+        effects: [setSignatureHelpInformationEffect.of({ pos, result })],
+      });
+    } catch (e) {
+      logException(this.view.state, e, "signature-help");
+      this.view.dispatch({
+        effects: [setSignatureHelpInformationEffect.of({ pos, result: null })],
+      });
+    }
+  }
+}
+
+function updateSignatureHelpState(
+  state: SignatureHelpState,
+  effect: SignatureChangeEffect
+): SignatureHelpState {
+  // Triggered by typing an opening bracket (may autocomplete the trailing one).
+  // Then persists so long as we get the same signature info back from
+  // the language server. If it changes (commonly but not always to null) then we stop.
+  // If the user types a new opening bracket then we restart the process.
+  if (!state.result && effect.result) {
+    const result = effect.result!;
+    return {
+      pos: effect.pos,
+      result,
+      tooltip: {
+        pos: effect.pos,
+        above: true,
+        create: () => {
+          let dom = document.createElement("div");
+          dom.className = "cm-cursor-tooltip";
+          dom.textContent = formatSignatureHelp(result);
+          return { dom };
+        },
+      },
+    };
+  } else if (
+    state.tooltip &&
+    ((state.result && !effect.result) ||
+      !isSameSignature(state.result!, effect.result!))
+  ) {
+    return {
+      pos: effect.pos,
+      result: null,
+      tooltip: null,
+    };
+  }
+  return state;
+}
+
+const formatSignatureHelp = (result: SignatureHelp): string => {
+  const { documentation, label } = result.signatures[result.activeSignature!];
+  if (documentation) {
+    if (MarkupContent.is(documentation)) {
+      return documentation.value;
+    }
+    return documentation;
+  }
+  return label;
+};
```

Comment:
This needs to be replaced with a DOM structure, ideally one that allows for interactions..

Reply:
Also, when is activeSignature not set? When it can't guess by params so far? So always for > 1 at point of completion?

I guess we'd have to show them all or default them and allow paging?

Reply:
Parking this for now. I want to merge what we have and address how we present it once we've got content we're happy with later this week.

All behind a flag.